core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2345,17 +2345,19 @@ private[spark] class RedirectThread(
  */
 private[spark] class CircularBuffer(sizeInBytes: Int = 10240) extends java.io.OutputStream {
   var pos: Int = 0
+  var isBufferFull = false
   var buffer = new Array[Int](sizeInBytes)
 
   def write(i: Int): Unit = {
     buffer(pos) = i
     pos = (pos + 1) % buffer.length
+    isBufferFull = isBufferFull || (pos == 0)
   }
 
   override def toString: String = {
     val (end, start) = buffer.splitAt(pos)
     val input = new java.io.InputStream {
-      val iterator = (start ++ end).iterator
+      val iterator = (if (isBufferFull) start ++ end else end).iterator
```

Comment:
This also becomes fairly inefficient, since it copies the entire buffer to make the two halves.
Since we're already copying and trying to parse this as a string, rather than make an input stream that reads byte by byte, how about just converting one or both halves of the byte array to a string and reading it with `StringReader`?